photonpump/discovery.py

```diff
@@ -212,8 +212,16 @@ class SingleNodeDiscovery:
 
     def __init__(self, node):
         self.node = node
+        self.failed = False
+
+    def mark_failed(self, node):
+        if node == self.node:
+            self.failed = True
 
     async def discover(self):
+        if self.failed:
+            raise DiscoveryFailed()
+        LOG.debug("SingleNodeDiscovery returning node %s", self.node)
```

Comment:
minor thing:
do we want to use `%s` notation or `f-strings`?

Reply:
For logs, I think boring old format. IF you use an f-string, you force interpolation even if that log level isn't enabled. For photonpump, which logs a _lot_ at trace and debug levels, that's a whole lot of string interpolation that we can avoid in performance-critical sections of code.